File: solutions/2026-07-04/java/2026-07-04-minimum-score-of-a-path-between-two-cities.java
```java
// Problem Summary: Find the minimum score of a path between city 1 and city n, where the score is the minimum edge weight in the path.
// Problem Link: https://leetcode.com/problems/minimum-score-of-a-path-between-two-cities/
//
// Approach:
// The problem asks for the minimum score of a path between city 1 and city n. The score of a path is the minimum edge weight within that path.
// Since we can traverse roads multiple times and visit cities multiple times, any road connected to city 1 or city n, or any city reachable from city 1,
// can potentially be part of a path that ultimately reaches city n.
// The key insight is that if city 1 and city n are in the same connected component, the minimum score of a path between them will be the minimum edge weight
// among ALL edges that are part of the connected component containing city 1 and city n. This is because we can construct a path that includes any edge in this component.
// For example, to include an edge (u, v) with weight w in our path from 1 to n (assuming 1 and n are in the same component as u and v), we can travel from 1 to u,
// traverse the edge (u, v), and then travel from v to n. The score of this path would be min(..., w, ...). If we can always find a path from 1 to n, and w is the smallest edge
// in the component, we can ensure w is the minimum edge by constructing a path that includes it.
//
// We can use either Breadth-First Search (BFS) or Depth-First Search (DFS) to explore the connected component starting from city 1.
// During the traversal, we keep track of the minimum edge weight encountered so far. Since we are guaranteed that city n is reachable from city 1,
// we will eventually explore all edges within their connected component.
//
// Algorithm using BFS:
// 1. Build an adjacency list representation of the graph where each entry stores the neighbor and the distance to that neighbor.
// 2. Initialize a queue for BFS and a boolean array `visited` to keep track of visited cities.
// 3. Initialize `minScore` to infinity.
// 4. Add city 1 to the queue and mark it as visited.
// 5. While the queue is not empty:
//    a. Dequeue a city `currentCity`.
//    b. For each neighbor `neighbor` of `currentCity` with edge weight `distance`:
//       i. Update `minScore = min(minScore, distance)`.
//       ii. If `neighbor` has not been visited:
//           - Mark `neighbor` as visited.
//           - Enqueue `neighbor`.
// 6. Return `minScore`.
//
// Algorithm using DFS:
// 1. Build an adjacency list representation of the graph.
// 2. Initialize a boolean array `visited` and `minScore` to infinity.
// 3. Call a recursive DFS function starting from city 1.
//
// DFS function `dfs(city, adj, visited, minScore)`:
// 1. Mark `city` as visited.
// 2. For each neighbor `neighbor` and `distance` of `city`:
//    a. Update `minScore = min(minScore, distance)`.
//    b. If `neighbor` is not visited:
//       i. Recursively call `dfs(neighbor, adj, visited, minScore)`.
//
// Time Complexity: O(V + E), where V is the number of cities and E is the number of roads. We visit each city and edge in the connected component at most once.
// Space Complexity: O(V + E) for the adjacency list. O(V) for the visited array and the recursion stack (for DFS) or the queue (for BFS).
//
// Alternative using Union-Find:
// While Union-Find can identify connected components, directly using it to find the minimum edge weight within a component can be slightly more complex.
// You would first union all cities connected by roads. Then, you would iterate through all roads and check if both endpoints of a road belong to the same component as city 1 and city n.
// If they do, update the minimum score. This approach might involve iterating through roads multiple times or storing edge information with components.
// The BFS/DFS approach is more direct for this specific problem.
//
// Given the constraints and the nature of the problem (finding the minimum edge in a connected component), BFS is a straightforward and efficient solution.
// We will implement the BFS approach.

import java.util.*;

class Solution {
    /**
     * Finds the minimum score of a path between city 1 and city n.
     * The score of a path is the minimum distance of a road in that path.
     *
     * @param n     The number of cities.
     * @param roads A 2D array where roads[i] = [ai, bi, distancei] indicates a bidirectional road.
     * @return The minimum possible score of a path between city 1 and city n.
     */
    public int minScore(int n, int[][] roads) {
        // Adjacency list to represent the graph.
        // Map: City -> List of Pairs (Neighbor City, Distance)
        Map<Integer, List<int[]>> adj = new HashMap<>();

        // Build the adjacency list from the roads.
        // Since roads are bidirectional, add entries for both directions.
        for (int[] road : roads) {
            int u = road[0];
            int v = road[1];
            int distance = road[2];

            adj.computeIfAbsent(u, k -> new ArrayList<>()).add(new int[]{v, distance});
            adj.computeIfAbsent(v, k -> new ArrayList<>()).add(new int[]{u, distance});
        }

        // Variable to store the minimum score found so far.
        // Initialize to a very large value.
        int minScore = Integer.MAX_VALUE;

        // Queue for Breadth-First Search (BFS).
        Queue<Integer> queue = new LinkedList<>();
        // Set to keep track of visited cities during BFS.
        Set<Integer> visited = new HashSet<>();

        // Start BFS from city 1.
        queue.offer(1);
        visited.add(1);

        // Perform BFS traversal.
        while (!queue.isEmpty()) {
            int currentCity = queue.poll();

            // If the current city is not in the adjacency list (e.g., isolated city, though problem guarantees reachability),
            // we can skip processing its neighbors.
            if (!adj.containsKey(currentCity)) {
                continue;
            }

            // Iterate through all roads connected to the current city.
            for (int[] neighborInfo : adj.get(currentCity)) {
                int neighborCity = neighborInfo[0];
                int distance = neighborInfo[1];

                // Update the minimum score encountered. The score of any path between 1 and n within this connected component
                // can be at most the minimum edge weight within this component.
                minScore = Math.min(minScore, distance);

                // If the neighbor city has not been visited yet, add it to the queue and mark it as visited.
                // This ensures we explore the entire connected component that includes city 1 and city n.
                if (!visited.contains(neighborCity)) {
                    visited.add(neighborCity);
                    queue.offer(neighborCity);
                }
            }
        }

        // After BFS, minScore will hold the minimum edge weight found in the connected component
        // containing city 1 and city n. This is because any path between 1 and n must traverse
        // through this connected component, and we can construct a path that utilizes any edge
        // within this component.
        return minScore;
    }
}
```

File: solutions/2026-07-04/python/2026-07-04-minimum-score-of-a-path-between-two-cities.py
```python
# Problem Summary: Find the minimum road distance among all roads on any path between city 1 and city n.
# Link: https://leetcode.com/problems/minimum-score-of-a-path-between-two-cities/
#
# Approach:
# The problem asks for the minimum score of a path between city 1 and city n. The score of a path is the minimum distance of any road in that path.
# Since we can traverse roads multiple times and revisit cities, any road connected to city 1 or city n, or any road on *any* path connecting city 1 to city n, can be part of an optimal path.
# This implies that if city 1 and city n are in the same connected component, the minimum score of a path between them will be the minimum distance of *any* road within that connected component.
#
# We can use Breadth-First Search (BFS) or Depth-First Search (DFS) starting from city 1 to explore all reachable cities and roads in its connected component. During the traversal, we keep track of the minimum road distance encountered so far.
#
# Alternatively, Union-Find could be used to group cities into connected components. After processing all roads, we can find the component containing city 1 and then iterate through all roads to find the minimum distance of roads that connect two cities within this component. However, a simpler BFS/DFS approach directly finds the minimum in the relevant component.
#
# The BFS/DFS approach is more straightforward:
# 1. Build an adjacency list representation of the graph where each entry stores the neighbor city and the road distance.
# 2. Initialize a `visited` set to keep track of visited cities and a `min_score` to infinity.
# 3. Start a BFS (or DFS) from city 1.
# 4. Use a queue for BFS. Add city 1 to the queue and `visited` set.
# 5. While the queue is not empty:
#    a. Dequeue a `current_city`.
#    b. For each `neighbor_city` and `distance` connected to `current_city`:
#       i. Update `min_score = min(min_score, distance)`.
#       ii. If `neighbor_city` has not been visited, add it to the queue and `visited` set.
# 6. After the BFS completes, `min_score` will hold the minimum road distance in the connected component containing city 1. Since city n is guaranteed to be reachable from city 1, this `min_score` is the answer.
#
# Time Complexity: O(V + E), where V is the number of cities (n) and E is the number of roads. This is because we visit each city and edge at most once during the BFS traversal.
# Space Complexity: O(V + E) for the adjacency list and O(V) for the visited set and queue in the worst case.
class Solution:
    def minScore(self, n: int, roads: list[list[int]]) -> int:
        # Adjacency list to represent the graph.
        # adj[city] will store a list of tuples (neighbor_city, distance).
        adj = [[] for _ in range(n + 1)]
        for u, v, d in roads:
            adj[u].append((v, d))
            adj[v].append((u, d))

        # Initialize the minimum score to a very large value.
        min_score = float('inf')

        # Use a set to keep track of visited cities to avoid cycles and redundant processing.
        visited = set()

        # Use a queue for Breadth-First Search (BFS).
        # Start BFS from city 1.
        queue = [1]
        visited.add(1)

        # Perform BFS traversal.
        while queue:
            current_city = queue.pop(0) # Dequeue the current city.

            # Iterate through all neighbors of the current city.
            for neighbor_city, distance in adj[current_city]:
                # Update the minimum score with the current road's distance.
                min_score = min(min_score, distance)

                # If the neighbor city has not been visited, add it to the queue and mark as visited.
                if neighbor_city not in visited:
                    visited.add(neighbor_city)
                    queue.append(neighbor_city)

        # The min_score after traversing the connected component of city 1
        # will be the minimum score of any path between city 1 and city n,
        # as any road in this component is accessible and can contribute to a path.
        return min_score

```